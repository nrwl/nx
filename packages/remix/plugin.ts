export {
  createNodes,
  createNodesV2,
  createDependencies,
  RemixPluginOptions,
} from './src/plugins/plugin';
