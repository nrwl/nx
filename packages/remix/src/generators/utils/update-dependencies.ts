import { type Tree, addDependenciesToPackageJson } from '@nx/devkit';
import {
  eslintVersion,
  isbotVersion,
  nxVersion,
  reactDomVersion,
  reactVersion,
  remixVersion,
  typescriptVersion,
  typesReactDomVersion,
  typesReactVersion,
  viteVersion,
} from '../../utils/versions';

export function updateDependencies(tree: Tree) {
  return addDependenciesToPackageJson(
    tree,
    {
      '@remix-run/node': remixVersion,
      '@remix-run/react': remixVersion,
      isbot: isbotVersion,
      react: reactVersion,
      'react-dom': reactDomVersion,
    },
    {
      '@types/react': typesReactVersion,
      '@types/react-dom': typesReactDomVersion,
      eslint: eslintVersion,
      typescript: typescriptVersion,
      vite: viteVersion,
      '@nx/vite': nxVersion,
    }
  );
}
