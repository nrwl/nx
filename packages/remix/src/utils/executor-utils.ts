export function loadViteDynamicImport() {
  return Function('return import("vite")')() as Promise<typeof import('vite')>;
}
