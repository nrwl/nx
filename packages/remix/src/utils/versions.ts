import { readJson, Tree } from '@nx/devkit';

export const nxVersion = require('../../package.json').version;

export const remixVersion = '^2.14.0';
export const isbotVersion = '^4.4.0';
export const reactVersion = '^18.2.0';
export const reactDomVersion = '^18.2.0';
export const typesReactVersion = '^18.2.0';
export const typesReactDomVersion = '^18.2.0';
export const eslintVersion = '^8.56.0';
export const typescriptVersion = '~5.6.2';
export const tailwindVersion = '^3.3.0';
export const postcssVersion = '^8.4.38';
export const autoprefixerVersion = '^10.4.19';
export const testingLibraryReactVersion = '^14.1.2';
// TODO(colum): Unpin this when @testing-library/jest-dom pushes a fix
export const testingLibraryJestDomVersion = '6.4.2';
export const testingLibraryUserEventsVersion = '^14.5.2';
export const viteVersion = '^5.0.0';

export function getRemixVersion(tree: Tree): string {
  return getPackageVersion(tree, '@remix-run/dev') ?? remixVersion;
}

export function getPackageVersion(tree: Tree, packageName: string) {
  const packageJsonContents = readJson(tree, 'package.json');
  return (
    packageJsonContents?.['devDependencies']?.[packageName] ??
    packageJsonContents?.['dependencies']?.[packageName] ??
    null
  );
}
