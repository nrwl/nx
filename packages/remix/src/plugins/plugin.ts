import { workspaceDataDirectory } from 'nx/src/utils/cache-directory';
import { hashObject } from 'nx/src/hasher/file-hasher';
import {
  type CreateDependencies,
  type CreateNodes,
  type CreateNodesContext,
  createNodesFromFiles,
  CreateNodesV2,
  detectPackageManager,
  joinPathFragments,
  logger,
  ProjectConfiguration,
  readJsonFile,
  type TargetConfiguration,
  writeJsonFile,
} from '@nx/devkit';
import { calculateHashForCreateNodes } from '@nx/devkit/src/utils/calculate-hash-for-create-nodes';
import { getNamedInputs } from '@nx/devkit/src/utils/get-named-inputs';
import { loadConfigFile } from '@nx/devkit/src/utils/config-utils';
import { getLockFileName } from '@nx/js';
import { type AppConfig } from '@remix-run/dev';
import { dirname, isAbsolute, join, relative } from 'path';
import { existsSync, readdirSync, readFileSync } from 'fs';
import { loadViteDynamicImport } from '../utils/executor-utils';

export interface RemixPluginOptions {
  buildTargetName?: string;
  devTargetName?: string;
  startTargetName?: string;
  typecheckTargetName?: string;
  /**
   * @deprecated Use serveStaticTargetName instead. This option will be removed in Nx 21.
   */
  staticServeTargetName?: string;
  serveStaticTargetName?: string;
}

type RemixTargets = Pick<ProjectConfiguration, 'targets' | 'metadata'>;

function readTargetsCache(
  cachePath: string
): Record<string, Record<string, TargetConfiguration>> {
  return existsSync(cachePath) ? readJsonFile(cachePath) : {};
}

function writeTargetsToCache(
  cachePath: string,
  results: Record<string, RemixTargets>
) {
  writeJsonFile(cachePath, results);
}

/**
 * @deprecated The 'createDependencies' function is now a no-op. This functionality is included in 'createNodesV2'.
 */
export const createDependencies: CreateDependencies = () => {
  return [];
};

const remixConfigGlob = '**/{remix,vite}.config.{js,cjs,mjs,ts,cts,mts}';

export const createNodesV2: CreateNodesV2<RemixPluginOptions> = [
  remixConfigGlob,
  async (configFilePaths, options, context) => {
    const optionsHash = hashObject(options);
    const cachePath = join(workspaceDataDirectory, `remix-${optionsHash}.hash`);
    const targetsCache = readTargetsCache(cachePath);
    try {
      return await createNodesFromFiles(
        (configFile, options, context) =>
          createNodesInternal(configFile, options, context, targetsCache),
        configFilePaths,
        options,
        context
      );
    } finally {
      writeTargetsToCache(cachePath, targetsCache);
    }
  },
];

export const createNodes: CreateNodes<RemixPluginOptions> = [
  remixConfigGlob,
  async (configFilePath, options, context) => {
    logger.warn(
      '`createNodes` is deprecated. Update your plugin to utilize createNodesV2 instead. In Nx 20, this will change to the createNodesV2 API.'
    );
    return createNodesInternal(configFilePath, options, context, {});
  },
];

async function createNodesInternal(
  configFilePath: string,
  options: RemixPluginOptions,
  context: CreateNodesContext,
  targetsCache: Record<string, RemixTargets>
) {
  const projectRoot = dirname(configFilePath);
  const fullyQualifiedProjectRoot = join(context.workspaceRoot, projectRoot);
  // Do not create a project if package.json and project.json isn't there
  const siblingFiles = readdirSync(fullyQualifiedProjectRoot);
  if (
    !siblingFiles.includes('package.json') &&
    !siblingFiles.includes('project.json')
  ) {
    return {};
  }

  options = normalizeOptions(options);

  const remixCompiler = determineIsRemixVite(
    configFilePath,
    context.workspaceRoot
  );

  if (remixCompiler === RemixCompiler.IsNotRemix) {
    return {};
  }

  const hash =
    (await calculateHashForCreateNodes(projectRoot, options, context, [
      getLockFileName(detectPackageManager(context.workspaceRoot)),
    ])) + configFilePath;

  targetsCache[hash] ??= await buildRemixTargets(
    configFilePath,
    projectRoot,
    options,
    context,
    siblingFiles,
    remixCompiler
  );

  const { targets, metadata } = targetsCache[hash];

  const project: ProjectConfiguration = {
    root: projectRoot,
    targets,
    metadata,
  };

  return {
    projects: {
      [projectRoot]: project,
    },
  };
}

async function buildRemixTargets(
  configFilePath: string,
  projectRoot: string,
  options: RemixPluginOptions,
  context: CreateNodesContext,
  siblingFiles: string[],
  remixCompiler: RemixCompiler
) {
  const namedInputs = getNamedInputs(projectRoot, context);
  const { buildDirectory, assetsBuildDirectory, serverBuildPath } =
    await getBuildPaths(
      configFilePath,
      projectRoot,
      context.workspaceRoot,
      remixCompiler
    );

  const targets: Record<string, TargetConfiguration> = {};
  targets[options.buildTargetName] = buildTarget(
    options.buildTargetName,
    projectRoot,
    buildDirectory,
    assetsBuildDirectory,
    namedInputs,
    remixCompiler
  );
  targets[options.devTargetName] = devTarget(
    serverBuildPath,
    projectRoot,
    remixCompiler
  );
  targets[options.startTargetName] = startTarget(
    projectRoot,
    serverBuildPath,
    options.buildTargetName,
    remixCompiler
  );
  // TODO(colum): Remove for Nx 21
  targets[options.staticServeTargetName] = startTarget(
    projectRoot,
    serverBuildPath,
    options.buildTargetName,
    remixCompiler
  );
  targets[options.serveStaticTargetName] = startTarget(
    projectRoot,
    serverBuildPath,
    options.buildTargetName,
    remixCompiler
  );
  targets[options.typecheckTargetName] = typecheckTarget(
    projectRoot,
    namedInputs,
    siblingFiles
  );

  return { targets, metadata: {} };
}

function buildTarget(
  buildTargetName: string,
  projectRoot: string,
  buildDirectory: string,
  assetsBuildDirectory: string,
  namedInputs: { [inputName: string]: any[] },
  remixCompiler: RemixCompiler
): TargetConfiguration {
  const serverBuildOutputPath =
    projectRoot === '.'
      ? joinPathFragments(`{workspaceRoot}`, buildDirectory)
      : joinPathFragments(`{workspaceRoot}`, projectRoot, buildDirectory);

  const assetsBuildOutputPath =
    projectRoot === '.'
      ? joinPathFragments(`{workspaceRoot}`, assetsBuildDirectory)
      : joinPathFragments(`{workspaceRoot}`, projectRoot, assetsBuildDirectory);

  const outputs =
    remixCompiler === RemixCompiler.IsVte
      ? [
          projectRoot === '.'
            ? joinPathFragments(`{workspaceRoot}`, buildDirectory)
            : joinPathFragments(`{workspaceRoot}`, projectRoot, buildDirectory),
        ]
      : [serverBuildOutputPath, assetsBuildOutputPath];

  return {
    cache: true,
    dependsOn: [`^${buildTargetName}`],
    inputs: [
      ...('production' in namedInputs
        ? ['production', '^production']
        : ['default', '^default']),
      { externalDependencies: ['@remix-run/dev'] },
    ],
    outputs,
    command:
      remixCompiler === RemixCompiler.IsVte
        ? 'remix vite:build'
        : 'remix build',
    options: { cwd: projectRoot },
  };
}

function devTarget(
  serverBuildPath: string,
  projectRoot: string,
  remixCompiler: RemixCompiler
): TargetConfiguration {
  return {
    command:
      remixCompiler === RemixCompiler.IsVte
        ? 'remix vite:dev'
        : 'remix dev --manual',
    options: { cwd: projectRoot },
  };
}

function startTarget(
  projectRoot: string,
  serverBuildPath: string,
  buildTargetName: string,
  remixCompiler: RemixCompiler
): TargetConfiguration {
  let serverPath = serverBuildPath;
  if (remixCompiler === RemixCompiler.IsVte) {
    if (serverBuildPath === 'build') {
      serverPath = `${serverBuildPath}/server/index.js`;
    }
  }

  return {
    dependsOn: [buildTargetName],
    command: `remix-serve ${serverPath}`,
    options: {
      cwd: projectRoot,
    },
  };
}

function typecheckTarget(
  projectRoot: string,
  namedInputs: { [inputName: string]: any[] },
  siblingFiles: string[]
): TargetConfiguration {
  const hasTsConfigAppJson = siblingFiles.includes('tsconfig.app.json');
  const command = `tsc${
    hasTsConfigAppJson ? ` --project tsconfig.app.json` : ``
  }`;
  return {
    command,
    cache: true,
    inputs: [
      ...('production' in namedInputs
        ? ['production', '^production']
        : ['default', '^default']),
      { externalDependencies: ['typescript'] },
    ],
    options: {
      cwd: projectRoot,
    },
  };
}

async function getBuildPaths(
  configFilePath: string,
  projectRoot: string,
  workspaceRoot: string,
  remixCompiler: RemixCompiler
): Promise<{
  buildDirectory: string;
  assetsBuildDirectory?: string;
  serverBuildPath?: string;
}> {
  const configPath = join(workspaceRoot, configFilePath);
  if (remixCompiler === RemixCompiler.IsClassic) {
    let appConfig = await loadConfigFile<AppConfig>(configPath);
    return {
      buildDirectory: 'build',
      serverBuildPath: appConfig.serverBuildPath ?? 'build/index.js',
      assetsBuildDirectory: appConfig.assetsBuildDirectory ?? 'public/build',
    };
  } else {
    // Workaround for the `build$3 is not a function` error that we sometimes see in agents.
    // This should be removed later once we address the issue properly
    try {
      const importEsbuild = () => new Function('return import("esbuild")')();
      await importEsbuild();
    } catch {
      // do nothing
    }
    const { resolveConfig } = await loadViteDynamicImport();
    const viteBuildConfig = (await resolveConfig(
      {
        configFile: configPath,
        mode: 'development',
      },
      'build'
    )) as any;

    return {
      buildDirectory: viteBuildConfig.build?.outDir ?? 'build',
      serverBuildPath: viteBuildConfig.build?.outDir
        ? join(
            dirname(viteBuildConfig.build?.outDir),
            `server/${viteBuildConfig.__remixPluginContext?.remixConfig.serverBuildFile}`
          )
        : 'build',
      assetsBuildDirectory: 'build/client',
    };
  }
}

function normalizeOptions(options: RemixPluginOptions) {
  options ??= {};
  options.buildTargetName ??= 'build';
  options.devTargetName ??= 'dev';
  options.startTargetName ??= 'start';
  options.typecheckTargetName ??= 'typecheck';
  // TODO(colum): remove for Nx 21
  options.staticServeTargetName ??= 'static-serve';
  options.serveStaticTargetName ??= 'serve-static';

  return options;
}

function determineIsRemixVite(configFilePath: string, workspaceRoot: string) {
  if (configFilePath.includes('remix.config')) {
    return RemixCompiler.IsClassic;
  }

  const VITE_PLUGIN_REGEX = /vitePlugin\(\s*(.|\n)*?\s*\)/;
  const REMIX_PLUGIN_REGEX = /remix\(\s*(.|\n)*?\s*\)/;

  const fileContents = readFileSync(
    join(workspaceRoot, configFilePath),
    'utf8'
  );
  if (
    fileContents.includes('@remix-run/dev') &&
    (VITE_PLUGIN_REGEX.test(fileContents) ||
      REMIX_PLUGIN_REGEX.test(fileContents))
  ) {
    return RemixCompiler.IsVte;
  } else {
    return RemixCompiler.IsNotRemix;
  }
}

enum RemixCompiler {
  IsClassic = 1,
  IsVte = 2,
  IsNotRemix = 3,
}
