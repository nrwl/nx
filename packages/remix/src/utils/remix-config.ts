import {
  joinPathFragments,
  readProjectConfiguration,
  Tree,
  workspaceRoot,
} from '@nx/devkit';
import type { AppConfig } from '@remix-run/dev';
import { loadViteDynamicImport } from './executor-utils';

export function getRemixConfigPathDetails(tree: Tree, projectName: string) {
  const project = readProjectConfiguration(tree, projectName);
  if (!project) throw new Error(`Project does not exist: ${projectName}`);

  for (const ext of ['.mjs', '.cjs', '.js', '.mts', '.cts', '.ts']) {
    const configPath = joinPathFragments(project.root, `vite.config${ext}`);
    if (tree.exists(configPath)) {
      return [configPath, 'vite'];
    }
  }

  for (const ext of ['.mjs', '.cjs', '.js']) {
    const configPath = joinPathFragments(project.root, `remix.config${ext}`);
    if (tree.exists(configPath)) {
      return [configPath, 'classic'];
    }
  }
}

export function getRemixConfigPath(tree: Tree, projectName: string) {
  const project = readProjectConfiguration(tree, projectName);
  if (!project) throw new Error(`Project does not exist: ${projectName}`);

  for (const ext of ['.mjs', '.cjs', '.js']) {
    const configPath = joinPathFragments(project.root, `remix.config${ext}`);
    if (tree.exists(configPath)) {
      return configPath;
    }
  }
}

export function getRemixConfigPathFromProjectRoot(
  tree: Tree,
  projectRoot: string
) {
  let pathToRemixConfig: string;
  for (const ext of ['.js', '.mjs', '.cjs']) {
    pathToRemixConfig = joinPathFragments(projectRoot, `remix.config${ext}`);
    if (tree.exists(pathToRemixConfig)) {
      return pathToRemixConfig;
    }
  }

  throw new Error(
    `Could not find a Remix Config File. Please ensure a "remix.config.js" file exists at the root of your project.`
  );
}

const _remixConfigCache: Record<string, AppConfig> = {};

export async function getRemixConfigValues(tree: Tree, projectName: string) {
  const [configPath, configType] = getRemixConfigPathDetails(tree, projectName);
  const remixConfigPath = joinPathFragments(workspaceRoot, configPath);
  const cacheKey = `${projectName}/${remixConfigPath}`;
  let appConfig = _remixConfigCache[cacheKey];

  let resolvedConfig: any;
  if (!appConfig) {
    if (configType === 'vite') {
      const { resolveConfig } = await loadViteDynamicImport();
      const viteBuildConfig = (await resolveConfig(
        {
          configFile: configPath,
          mode: 'development',
        },
        'build'
      )) as any;
      appConfig = viteBuildConfig.__remixPluginContext?.remixConfig;
    } else {
      try {
        const importedConfig = await Function(
          `return import("${remixConfigPath}?t=${Date.now()}")`
        )();
        appConfig = (importedConfig?.default || importedConfig) as AppConfig;
      } catch {
        appConfig = require(remixConfigPath);
      }
    }
    _remixConfigCache[cacheKey] = appConfig;
  }

  return appConfig;
}
