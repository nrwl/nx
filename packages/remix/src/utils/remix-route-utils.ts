import { joinPathFragments, readProjectConfiguration, Tree } from '@nx/devkit';
import { relative } from 'path';
import { getRemixConfigValues } from './remix-config';

export function normalizeRoutePath(path: string) {
  return path.indexOf('/routes/') > -1
    ? path.substring(path.indexOf('/routes/') + 8)
    : path;
}

export function checkRoutePathForErrors(path: string) {
  return (
    path.match(/\w\.\.\w/) || // route.$withParams.tsx => route..tsx
    path.match(/\w\/\/\w/) || // route/$withParams/index.tsx => route//index.tsx
    path.match(/\w\/\.\w/) // route/$withParams.tsx => route/.tsx
  );
}

export async function resolveRemixAppDirectory(
  tree: Tree,
  projectName: string
) {
  const project = readProjectConfiguration(tree, projectName);
  const remixConfig = await getRemixConfigValues(tree, projectName);

  return joinPathFragments(
    project.root,
    remixConfig.appDirectory
      ? relative(project.root, remixConfig.appDirectory)
      : 'app'
  );
}
