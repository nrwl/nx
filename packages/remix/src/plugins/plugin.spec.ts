import { type CreateNodesContext, joinPathFragments } from '@nx/devkit';
import { createNodesV2 as createNodes } from './plugin';
import { TempFs } from 'nx/src/internal-testing-utils/temp-fs';
import { loadViteDynamicImport } from '../utils/executor-utils';

jest.mock('../utils/executor-utils', () => ({
  loadViteDynamicImport: jest.fn().mockResolvedValue({
    resolveConfig: jest.fn().mockResolvedValue({}),
  }),
}));

describe('@nx/remix/plugin', () => {
  let createNodesFunction = createNodes[1];
  let context: CreateNodesContext;
  let cwd = process.cwd();

  describe('Remix Classic Compiler', () => {
    describe('root project', () => {
      const tempFs = new TempFs('test');

      beforeEach(() => {
        context = {
          nxJsonConfiguration: {
            targetDefaults: {
              build: {
                cache: false,
                inputs: ['foo', '^foo'],
              },
              dev: {
                command: 'npm run dev',
              },
              start: {
                command: 'npm run start',
              },
              typecheck: {
                command: 'tsc',
              },
            },
            namedInputs: {
              default: ['{projectRoot}/**/*'],
              production: ['!{projectRoot}/**/*.spec.ts'],
            },
          },
          workspaceRoot: tempFs.tempDir,
          configFiles: [],
        };
        tempFs.createFileSync(
          'package.json',
          JSON.stringify('{name: "my-app", type: "module"}')
        );
        tempFs.createFileSync(
          'remix.config.cjs',
          `/**
 * @type {import('@remix-run/dev').AppConfig}
 */
module.exports = {
  ignoredRouteFiles: ['**/.*'],
  watchPaths: () => require('@nx/remix').createWatchPaths(__dirname),
};
`
        );
        process.chdir(tempFs.tempDir);
      });

      afterEach(() => {
        jest.resetModules();
        tempFs.cleanup();
        process.chdir(cwd);
      });

      it('should create nodes', async () => {
        // ACT
        const nodes = await createNodesFunction(
          ['remix.config.cjs'],
          {
            buildTargetName: 'build',
            devTargetName: 'dev',
            startTargetName: 'start',
            typecheckTargetName: 'typecheck',
            staticServeTargetName: 'static-serve',
          },
          context
        );

        // ASSERT
        expect(nodes).toMatchSnapshot();
      });
    });

    describe('non-root project', () => {
      const tempFs = new TempFs('test');

      beforeEach(() => {
        context = {
          nxJsonConfiguration: {
            namedInputs: {
              default: ['{projectRoot}/**/*'],
              production: ['!{projectRoot}/**/*.spec.ts'],
            },
          },
          workspaceRoot: tempFs.tempDir,
          configFiles: [],
        };

        tempFs.createFileSync(
          'my-app/project.json',
          JSON.stringify({ name: 'my-app' })
        );

        tempFs.createFileSync(
          'my-app/remix.config.cjs',
          `/**
 * @type {import('@remix-run/dev').AppConfig}
 */
module.exports = {
  ignoredRouteFiles: ['**/.*'],
  watchPaths: () => require('@nx/remix').createWatchPaths(__dirname),
};
`
        );

        process.chdir(tempFs.tempDir);
      });

      afterEach(() => {
        jest.resetModules();
        tempFs.cleanup();
        process.chdir(cwd);
      });

      it('should create nodes', async () => {
        // ACT
        const nodes = await createNodesFunction(
          ['my-app/remix.config.cjs'],
          {
            buildTargetName: 'build',
            devTargetName: 'dev',
            startTargetName: 'start',
            typecheckTargetName: 'tsc',
            staticServeTargetName: 'static-serve',
          },
          context
        );

        // ASSERT
        expect(nodes).toMatchSnapshot();
      });

      it('should infer watch-deps target', async () => {
        tempFs.createFileSync(
          'my-app/package.json',
          JSON.stringify('{"name": "my-app"}')
        );

        const nodes = await createNodesFunction(
          ['my-app/remix.config.cjs'],
          {
            buildTargetName: 'build',
            devTargetName: 'dev',
            startTargetName: 'start',
            typecheckTargetName: 'tsc',
            staticServeTargetName: 'static-serve',
          },
          context
        );

        expect(nodes).toMatchSnapshot();
      });
    });
  });

  describe('Remix Vite Compiler', () => {
    describe('root project', () => {
      const tempFs = new TempFs('test');

      beforeEach(() => {
        context = {
          nxJsonConfiguration: {
            targetDefaults: {
              build: {
                cache: false,
                inputs: ['foo', '^foo'],
              },
              dev: {
                command: 'npm run dev',
              },
              start: {
                command: 'npm run start',
              },
              typecheck: {
                command: 'tsc',
              },
            },
            namedInputs: {
              default: ['{projectRoot}/**/*'],
              production: ['!{projectRoot}/**/*.spec.ts'],
            },
          },
          workspaceRoot: tempFs.tempDir,
          configFiles: [],
        };
        tempFs.createFileSync(
          'package.json',
          JSON.stringify('{name: "my-app", type: "module"}')
        );
        tempFs.createFileSync(
          'vite.config.js',
          `const {defineConfig} = require('vite');
          const { vitePlugin: remix } = require('@remix-run/dev');
          module.exports = defineConfig({
             plugins:[remix()]
          });`
        );
        process.chdir(tempFs.tempDir);
        (loadViteDynamicImport as jest.Mock).mockResolvedValue({
          resolveConfig: jest.fn().mockResolvedValue({
            build: {
              lib: {
                entry: 'index.ts',
                name: 'my-app',
              },
            },
          }),
        });
      });

      afterEach(() => {
        jest.resetModules();
        tempFs.cleanup();
        process.chdir(cwd);
      });

      it('should create nodes', async () => {
        // ACT
        const nodes = await createNodesFunction(
          ['vite.config.js'],
          {
            buildTargetName: 'build',
            devTargetName: 'dev',
            startTargetName: 'start',
            typecheckTargetName: 'typecheck',
            staticServeTargetName: 'static-serve',
          },
          context
        );

        // ASSERT
        expect(nodes).toMatchSnapshot();
      });
    });

    describe('non-root project', () => {
      const tempFs = new TempFs('test');

      beforeEach(() => {
        context = {
          nxJsonConfiguration: {
            namedInputs: {
              default: ['{projectRoot}/**/*'],
              production: ['!{projectRoot}/**/*.spec.ts'],
            },
          },
          workspaceRoot: tempFs.tempDir,
          configFiles: [],
        };

        tempFs.createFileSync(
          'my-app/project.json',
          JSON.stringify({ name: 'my-app' })
        );

        tempFs.createFileSync(
          'my-app/vite.config.js',
          `const {defineConfig} = require('vite');
          const { vitePlugin: remix } = require('@remix-run/dev');
          module.exports = defineConfig({
             plugins:[remix()]
          });`
        );
        (loadViteDynamicImport as jest.Mock).mockResolvedValue({
          resolveConfig: jest.fn().mockResolvedValue({
            build: {
              lib: {
                entry: 'index.ts',
                name: 'my-app',
              },
            },
          }),
        });

        process.chdir(tempFs.tempDir);
      });

      afterEach(() => {
        jest.resetModules();
        tempFs.cleanup();
        process.chdir(cwd);
      });

      it('should create nodes', async () => {
        // ACT
        const nodes = await createNodesFunction(
          ['my-app/vite.config.js'],
          {
            buildTargetName: 'build',
            devTargetName: 'dev',
            startTargetName: 'start',
            typecheckTargetName: 'tsc',
            staticServeTargetName: 'static-serve',
          },
          context
        );

        // ASSERT
        expect(nodes).toMatchSnapshot();
      });
    });
  });
});
