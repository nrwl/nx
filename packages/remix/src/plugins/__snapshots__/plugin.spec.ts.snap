// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@nx/remix/plugin Remix Classic Compiler non-root project should create nodes 1`] = `
[
  [
    "my-app/remix.config.cjs",
    {
      "projects": {
        "my-app": {
          "metadata": {},
          "root": "my-app",
          "targets": {
            "build": {
              "cache": true,
              "command": "remix build",
              "dependsOn": [
                "^build",
              ],
              "inputs": [
                "production",
                "^production",
                {
                  "externalDependencies": [
                    "@remix-run/dev",
                  ],
                },
              ],
              "options": {
                "cwd": "my-app",
              },
              "outputs": [
                "{workspaceRoot}/my-app/build",
                "{workspaceRoot}/my-app/public/build",
              ],
            },
            "build-deps": {
              "dependsOn": [
                "^build",
              ],
            },
            "dev": {
              "command": "remix dev --manual",
              "options": {
                "cwd": "my-app",
              },
            },
            "serve-static": {
              "command": "remix-serve build/index.js",
              "dependsOn": [
                "build",
              ],
              "options": {
                "cwd": "my-app",
              },
            },
            "start": {
              "command": "remix-serve build/index.js",
              "dependsOn": [
                "build",
              ],
              "options": {
                "cwd": "my-app",
              },
            },
            "static-serve": {
              "command": "remix-serve build/index.js",
              "dependsOn": [
                "build",
              ],
              "options": {
                "cwd": "my-app",
              },
            },
            "tsc": {
              "cache": true,
              "command": "tsc",
              "inputs": [
                "production",
                "^production",
                {
                  "externalDependencies": [
                    "typescript",
                  ],
                },
              ],
              "options": {
                "cwd": "my-app",
              },
            },
            "watch-deps": {
              "command": "npx nx watch --projects my-app --includeDependentProjects -- npx nx build-deps my-app",
              "dependsOn": [
                "build-deps",
              ],
            },
          },
        },
      },
    },
  ],
]
`;

exports[`@nx/remix/plugin Remix Classic Compiler non-root project should infer watch-deps target 1`] = `
[
  [
    "my-app/remix.config.cjs",
    {
      "projects": {
        "my-app": {
          "metadata": {},
          "root": "my-app",
          "targets": {
            "build": {
              "cache": true,
              "command": "remix build",
              "dependsOn": [
                "^build",
              ],
              "inputs": [
                "production",
                "^production",
                {
                  "externalDependencies": [
                    "@remix-run/dev",
                  ],
                },
              ],
              "options": {
                "cwd": "my-app",
              },
              "outputs": [
                "{workspaceRoot}/my-app/build",
                "{workspaceRoot}/my-app/public/build",
              ],
            },
            "build-deps": {
              "dependsOn": [
                "^build",
              ],
            },
            "dev": {
              "command": "remix dev --manual",
              "options": {
                "cwd": "my-app",
              },
            },
            "serve-static": {
              "command": "remix-serve build/index.js",
              "dependsOn": [
                "build",
              ],
              "options": {
                "cwd": "my-app",
              },
            },
            "start": {
              "command": "remix-serve build/index.js",
              "dependsOn": [
                "build",
              ],
              "options": {
                "cwd": "my-app",
              },
            },
            "static-serve": {
              "command": "remix-serve build/index.js",
              "dependsOn": [
                "build",
              ],
              "options": {
                "cwd": "my-app",
              },
            },
            "tsc": {
              "cache": true,
              "command": "tsc",
              "inputs": [
                "production",
                "^production",
                {
                  "externalDependencies": [
                    "typescript",
                  ],
                },
              ],
              "options": {
                "cwd": "my-app",
              },
            },
            "watch-deps": {
              "command": "npx nx watch --projects my-app --includeDependentProjects -- npx nx build-deps my-app",
              "dependsOn": [
                "build-deps",
              ],
            },
          },
        },
      },
    },
  ],
]
`;

exports[`@nx/remix/plugin Remix Classic Compiler root project should create nodes 1`] = `
[
  [
    "remix.config.cjs",
    {
      "projects": {
        ".": {
          "metadata": {},
          "root": ".",
          "targets": {
            "build": {
              "cache": true,
              "command": "remix build",
              "dependsOn": [
                "^build",
              ],
              "inputs": [
                "production",
                "^production",
                {
                  "externalDependencies": [
                    "@remix-run/dev",
                  ],
                },
              ],
              "options": {
                "cwd": ".",
              },
              "outputs": [
                "{workspaceRoot}/build",
                "{workspaceRoot}/public/build",
              ],
            },
            "dev": {
              "command": "remix dev --manual",
              "options": {
                "cwd": ".",
              },
            },
            "serve-static": {
              "command": "remix-serve build/index.js",
              "dependsOn": [
                "build",
              ],
              "options": {
                "cwd": ".",
              },
            },
            "start": {
              "command": "remix-serve build/index.js",
              "dependsOn": [
                "build",
              ],
              "options": {
                "cwd": ".",
              },
            },
            "static-serve": {
              "command": "remix-serve build/index.js",
              "dependsOn": [
                "build",
              ],
              "options": {
                "cwd": ".",
              },
            },
            "typecheck": {
              "cache": true,
              "command": "tsc",
              "inputs": [
                "production",
                "^production",
                {
                  "externalDependencies": [
                    "typescript",
                  ],
                },
              ],
              "options": {
                "cwd": ".",
              },
            },
          },
        },
      },
    },
  ],
]
`;

exports[`@nx/remix/plugin Remix Vite Compiler non-root project should create nodes 1`] = `
[
  [
    "my-app/vite.config.js",
    {
      "projects": {
        "my-app": {
          "metadata": {},
          "root": "my-app",
          "targets": {
            "build": {
              "cache": true,
              "command": "remix vite:build",
              "dependsOn": [
                "^build",
              ],
              "inputs": [
                "production",
                "^production",
                {
                  "externalDependencies": [
                    "@remix-run/dev",
                  ],
                },
              ],
              "options": {
                "cwd": "my-app",
              },
              "outputs": [
                "{workspaceRoot}/my-app/build",
              ],
            },
            "build-deps": {
              "dependsOn": [
                "^build",
              ],
            },
            "dev": {
              "command": "remix vite:dev",
              "options": {
                "cwd": "my-app",
              },
            },
            "serve-static": {
              "command": "remix-serve build/server/index.js",
              "dependsOn": [
                "build",
              ],
              "options": {
                "cwd": "my-app",
              },
            },
            "start": {
              "command": "remix-serve build/server/index.js",
              "dependsOn": [
                "build",
              ],
              "options": {
                "cwd": "my-app",
              },
            },
            "static-serve": {
              "command": "remix-serve build/server/index.js",
              "dependsOn": [
                "build",
              ],
              "options": {
                "cwd": "my-app",
              },
            },
            "tsc": {
              "cache": true,
              "command": "tsc",
              "inputs": [
                "production",
                "^production",
                {
                  "externalDependencies": [
                    "typescript",
                  ],
                },
              ],
              "options": {
                "cwd": "my-app",
              },
            },
            "watch-deps": {
              "command": "npx nx watch --projects my-app --includeDependentProjects -- npx nx build-deps my-app",
              "dependsOn": [
                "build-deps",
              ],
            },
          },
        },
      },
    },
  ],
]
`;

exports[`@nx/remix/plugin Remix Vite Compiler root project should create nodes 1`] = `
[
  [
    "vite.config.js",
    {
      "projects": {
        ".": {
          "metadata": {},
          "root": ".",
          "targets": {
            "build": {
              "cache": true,
              "command": "remix vite:build",
              "dependsOn": [
                "^build",
              ],
              "inputs": [
                "production",
                "^production",
                {
                  "externalDependencies": [
                    "@remix-run/dev",
                  ],
                },
              ],
              "options": {
                "cwd": ".",
              },
              "outputs": [
                "{workspaceRoot}/build",
              ],
            },
            "dev": {
              "command": "remix vite:dev",
              "options": {
                "cwd": ".",
              },
            },
            "serve-static": {
              "command": "remix-serve build/server/index.js",
              "dependsOn": [
                "build",
              ],
              "options": {
                "cwd": ".",
              },
            },
            "start": {
              "command": "remix-serve build/server/index.js",
              "dependsOn": [
                "build",
              ],
              "options": {
                "cwd": ".",
              },
            },
            "static-serve": {
              "command": "remix-serve build/server/index.js",
              "dependsOn": [
                "build",
              ],
              "options": {
                "cwd": ".",
              },
            },
            "typecheck": {
              "cache": true,
              "command": "tsc",
              "inputs": [
                "production",
                "^production",
                {
                  "externalDependencies": [
                    "typescript",
                  ],
                },
              ],
              "options": {
                "cwd": ".",
              },
            },
          },
        },
      },
    },
  ],
]
`;
